sbin/swapctl/swapctl.8
.\"	$NetBSD: swapctl.8,v 1.1 1997/06/12 13:14:11 mrg Exp $
.\"
.\" Copyright (c) 1997 Matthew R. Green
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed by Matthew R. Green.
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
.\" BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
.\" AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd June 12, 1997
.Dt SWAPCTL 8
.Os NetBSD
.Sh NAME
.Nm swapctl
.Nd system swap management tool
.Sh SYNOPSIS
.Nm swapctl
.Op Fl A
.Op Fl p Ar priority
.Nm swapctl
.Op Fl a | Fl c
.\" SWAP_OFF_WORKS: .Op Fl a | Fl c | Fl d
.Op Fl p Ar priority
.Ar path
.Nm swapctl
.Op Fl l | Fl s
.Op Fl k
.Nm swapon
.Op Fl a
.Ar path
.\" .Nm swapoff
.\" .Ar path
.Sh DESCRIPTION
The
.Nm
program adds,
.\" SWAP_OFF_WORKS: removes,
lists and prioritizes swap devices and files for the system.
The
.Nm swapon
program acts the same as the
.Nm
program, as if called with the
.Fl a
option, except if
.Nm swapon
itself is called with
.Fl a
in which case,
.Nm swapon
acts as 
.Nm
with the
.Fl A
option.
.Sh OPTIONS
The following options are available:
.Bl -tag -width 123456
.It Fl A
This options causes
.Nm
to read the
.Pa /etc/fstab
file for devices and files with a ``sw'' type, and adds all these entries
as swap devices.  The priority can set by having ``priority=N'' in the
mount options field of the 
.Xr fstab 5 
file.
.It Fl a
The
.Fl a
option requires that a
.Ar path
also be in the argument list.  The
.Ar path
is added to the kernel's list of swap devices using the
.Xr swapctl 2
system call.  When using the
.Nm swapon
form of this command, the
.Fl a
option is treated the same as the
.Fl A
option, for backwards compatibility.
.It Fl c
The
.Fl c
option changes the priority of the listed swap device or file.
.\" .It Fl d
.\" The
.\" .Fl d
.\" option removes the listed
.\" .Ar path
.\" from the kernel's list of swap devices or files.
.It Fl l
The
.Fl l
option lists the current swap devices and files, and their usage statistics.
.It Fl s
The
.Fl s
option displays a single line summary of current swap statistics.
.It Fl p
The
.Fl p
option sets the priority of swap devices or files to the
.Ar priority
argument.  This works with the
.\" .Fl d ,
.Fl a ,
.Fl c
and
.Fl l
options.
.It Fl k
The
.Fl k
option uses 1024 byte blocks instead of the default 512 byte.
.El
.Sh HISTORY
The
.Nm
program was first made available in
.Nx 1.3 .
The original
.Nm swapon
program appeared in
.Bx 4.0 ,
but is no longer distributed with
.Nx .
.Sh AUTHOR
The
.Nm
program was written by Matthew R. Green <mrg@eterna.com.au>.
.Sh SEE ALSO
.Xr swapctl 2 ,
.Xr fstab 5 .


sbin/swapctl/swapctl.c
/*	$NetBSD: swapctl.c,v 1.1 1997/06/12 13:14:11 mrg Exp $	*/

/*
 * Copyright (c) 1996, 1997 Matthew R. Green
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by Matthew R. Green for
 *      The NetBSD Foundation.
 * 4. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * swapctl command:
 *	-A		add all devices listed as `sw' in /etc/fstab
 *	-a <dev>	add this device
 *	-d <dev>	remove this swap device (not supported yet)
 *	-l		list swap devices
 *	-s		short listing of swap devices
 *	-k		use kilobytes
 *	-p <pri>	use this priority
 *	-c		change priority
 */

#include <sys/param.h>

#include <vm/vm_swap.h>

#include <unistd.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fstab.h>

#include "swapctl.h"

int	Aflag;
int	aflag;
int	cflag;
#ifdef SWAP_OFF_WORKS
int	dflag;
#endif
int	lflag;
int	kflag;
int	sflag;
int	pflag;
int	pri;		/* uses 0 as default pri */
char	*path;

static	void change_priority __P((char *));
static	void add_swap __P((char *));
#ifdef SWAP_OFF_WORKS
static	void del_swap __P((char *));
#endif /* SWAP_OFF_WORKS */
static	void do_fstab __P((void));
static	void usage __P((void));

int
main(argc, argv)
	int	argc;
	char	*argv[];
{
	int	c;
	int	am_swapon = 0;
#ifdef SWAP_OFF_WORKS
	int	swapoff = 0;
	static	char getoptstr[] = "Aacdlkp:s";
#else
	static	char getoptstr[] = "Aaclkp:s";
#endif /* SWAP_OFF_WORKS */
	extern	char *__progname;		/* XXX */

	if (strcmp(__progname, "swapon") == 0)
		aflag = am_swapon = 1;
#ifdef SWAP_OFF_WORKS
	else if (strcmp(__progname, "swapoff") == 0)
		dflag = swapoff = 1;
#endif /* SWAP_OFF_WORKS */
	while ((c = getopt(argc, argv, getoptstr)) != EOF) {
		switch(c) {
		case 'A':
			Aflag = 1;
			break;
		case 'a':
			if (cflag) {
				warn("-a and -c are mutually exclusive");
				usage();
			}
#ifdef SWAP_OFF_WORKS
			if (dflag) {
				warn("-a and -d are mutually exclusive");
				usage();
			}
#endif /* SWAP_OFF_WORKS */
			if (am_swapon)
				Aflag = 1;
			else
				aflag = 1;
			break;
		case 'c':
			if (aflag) {
				warn("-c and -a are mutually exclusive");
				usage();
			}
			cflag = 1;
			break;
#ifdef SWAP_OFF_WORKS
		case 'd':
			if (aflag || cflag) {
				warn("-d and -a or -c are mutually exclusive");
				usage();
			}
			dflag = 1;
			break;
#endif /* SWAP_OFF_WORKS */
		case 'l':
			lflag = 1;
			break;
		case 'k':
			kflag = 1;
			break;
		case 'p':
			pflag = 1;
			pri = atoi(optarg);
			break;
		case 's':
			sflag = 1;
			break;
		}
	}
	/* SWAP_OFF_WORKS */
	if (!aflag && !Aflag && !cflag && !lflag && !sflag /* && !dflag */)
		usage();

	argv += optind;
	if (!*argv && !Aflag && !lflag && !sflag)
		usage();
	if (cflag && !pflag)
		usage();

	if (lflag)
		list_swap(pri, kflag, pflag, 0, 1);
	else if (sflag)
		list_swap(pri, kflag, pflag, 0, 0);
	else if (cflag)
		change_priority(argv[0]);
	else if (aflag)
		add_swap(argv[0]);
#ifdef SWAP_OFF_WORKS
	else if (dflag)
		del_swap(argv[0]);
#endif /* SWAP_OFF_WORKS */
	else if (Aflag)
		do_fstab();
	exit(0);
}

/*
 * change_priority:  change the priority of a swap device.
 */
void
change_priority(path)
	char	*path;
{

	if (swapctl(SWAP_CTL, path, pri) < 0)
		warn("%s", path);
}

/*
 * add_swap:  add the pathname to the list of swap devices.
 */
void
add_swap(path)
	char *path;
{

	if (swapctl(SWAP_ON, path, pri) < 0)
		warn("%s", path);
}

#if SWAP_OFF_WORKS
/*
 * del_swap:  remove the pathname to the list of swap devices.
 *
 * XXX note that the kernel does not support this operation (yet).
 */
void
del_swap(path)
	char path;
{

	if (swapctl(SWAP_OFF, path, pri) < 0)
		warn("%s", path);
}
#endif /* SWAP_OFF_WORKS */

void
do_fstab()
{
	struct	fstab *fp;
	char	*s;
	long	priority;

#define PRIORITYEQ	"priority="
#define NFSMNTPT	"nfsmntpt="
#define PATH_MOUNT	"/sbin/mount_nfs"
	while (fp = getfsent()) {
		char *spec;

		if (strcmp(fp->fs_type, "sw") != 0)
			continue;

		spec = fp->fs_spec;

		if (s = strstr(fp->fs_mntops, PRIORITYEQ)) {
			s += sizeof(PRIORITYEQ) - 1;
			priority = atol(s);
		} else
			priority = pri;

		if (s = strstr(fp->fs_mntops, NFSMNTPT)) {
			char *t, cmd[2*PATH_MAX+sizeof(PATH_MOUNT)+2];

			t = strpbrk(s, ",");
			if (t != 0)
				*t = '\0';
			spec = strdup(s + strlen(NFSMNTPT));
			if (t != 0)
				*t = ',';

			if (spec == NULL)
				errx(1, "Out of memory");

			if (strlen(spec) == 0) {
				warnx("empty mountpoint");
				free(spec);
				continue;
			}
			snprintf(cmd, sizeof(cmd), "%s %s %s",
				PATH_MOUNT, fp->fs_spec, spec);
			if (system(cmd) != 0) {
				warnx("%s: mount failed", fp->fs_spec);
				continue;
			}
		}

		if (swapctl(SWAP_ON, spec, (int)priority) < 0)
			warn("%s", spec);
		else
printf("swap: adding %s as swap device at priority %d\n",
	    fp->fs_spec, priority);

		if (spec != fp->fs_spec)
			free(spec);
	}
}

void
usage()
{
	extern char *__progname;
#ifdef SWAP_OFF_WORKS
	static	char usagemsg[] =
	    "usage: %s [-k] [-A|-a|-c|-d|-l|-s] [-p <pri>] [device]\n";
#else
	static	char usagemsg[] =
	    "usage: %s [-k] [-A|-a|-c|-l|-s] [-p <pri>] [device]\n";
#endif /* SWAP_OFF_WORKS */

	fprintf(stderr, usagemsg, __progname);
	exit(1);
}
